### PR TITLE
ARGO-176 Enpoint group results

### PR DESCRIPTION
**Pull request by makistsantekidis**

### CODE REVIEW

**Comment by makistsantekidis** [comment]
please review

@pkoro @kaggis @skanct

File: app/results/Controller.go [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+	"labix.org/v2/mgo/bson"
+
+	"github.com/argoeu/argo-web-api/utils/authentication"
+	"github.com/argoeu/argo-web-api/utils/config"
+	"github.com/argoeu/argo-web-api/utils/mongo"
+	"github.com/gorilla/mux"
+)
+
+// THIS CONTROLLER IS JUST A DEMO AND IS NOT SOMETHING THAT WORKS.
+// TODO: WRITE AN ACTUAL CONTROLLER FOR AVAILABILITY
```
Comment (pkoro) [inline app/results/Controller.go]: Maybe remove these couple of lines.

File: app/results/Controller.go [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+	}
+
+	// Parse the request into the input
+	urlValues := r.URL.Query()
+	vars := mux.Vars(r)
+
+	input := endpointGroupResultQuery{
+		Name:        vars["lgroup_name"],
+		Granularity: urlValues.Get("granularity"),
+		Format:      strings.ToLower(urlValues.Get("format")),
```
Comment (pkoro) [inline app/results/Controller.go]: `format` should come in from headers, not url arguments.

Comment (pkoro) [inline app/results/Controller.go]: Hmm , I see it's correct a few lines below, so maybe just remove it from here and input struct?

File: app/results/Controller.go [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+
+	if err != nil {
+		code = http.StatusInternalServerError
+		return code, h, output, err
+	}
+
+	return code, h, output, err
+}
+
+func prepareFilter(input endpointGroupResultQuery) bson.M {
```
Comment (pkoro) [inline app/results/Controller.go]: Better remove this function entirely. You are preparing the filter within the `ListEndpointGroupResults` method above.

File: app/results/Controller.go (hunk 2 above)
Comment (makistsantekidis) [inline app/results/Controller.go]: Yep

File: app/results/Controller.go (hunk 3 above)
Comment (makistsantekidis) [inline app/results/Controller.go]: Yep forgot to do that. Thanks!

File: doc/api/results.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+## Endpoint Groups
+### Input
+Endpoint Groups
+
+```
+/results/{report_name}/{lesser_group_type}?[start_time]&[end_time]&[granularity]
+or
+/results/{report_name}/{lesser_group_type}/{lesser_group_name}?[start_time]&[end_time]&[granularity]
+or
+/results/{report_name}/{group_type}/{lesser_group_type}?[start_time]&[end_time]&[granularity]
```
Comment (pkoro) [inline doc/api/results.md]: Change this to

```
/results/{report_name}/{group_type}/{group_name}/{lesser_group_type}?[start_time]&[end_time]&[granularity]
```

File: doc/api/results.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+or
+/results/{report_name}/{group_type}/{lesser_group_type}/{lesser_group_name}?[start_time]&[end_time]&[granularity]
```
Comment (pkoro) [inline doc/api/results.md]: Change this to

```
/results/{report_name}/{group_type}/{group_name}/{lesser_group_type}/{lesser_group_name}?[start_time]&[end_time]&[granularity]
```

File: doc/api/results.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+`{group_name}`        | Name of the Endpoint Groups. If no name is specified then all Endpoint Groups are retrieved.              | NO       |
+`{group_type}`        | Name of the group that groups the Endpoint Groups. If no name is specified then all groups are retrieved. | NO       |
+`{lesser_group_type}` | Name of the group that groups the Endpoint Groups. If no name is specified then all groups are retrieved. | NO       |
+`{lesser_group_name}` | Name of the group that groups the Endpoint Groups. If no name is specified then all groups are retrieved. | NO       |
+
+#### Request headers
+
+```
+x-api-key: "tenant_key_value"
+format: "xml or json"
```
Comment (pkoro) [inline doc/api/results.md]: Not sure 100% but, according to our design plan this should be

```
Accept: application/xml
```

(or
```
Accept: application/json
```
I suspect in the case for json). 

In any case, this is not just something for the documentation. Maybe you need changes in the tests and actual code.

File: doc/api/results.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+/results/{report_name}/{lesser_group_type}?[start_time]&[end_time]&[granularity]
+or
+/results/{report_name}/{lesser_group_type}/{lesser_group_name}?[start_time]&[end_time]&[granularity]
+or
+/results/{report_name}/{group_type}/{lesser_group_type}?[start_time]&[end_time]&[granularity]
+or
+/results/{report_name}/{group_type}/{lesser_group_type}/{lesser_group_name}?[start_time]&[end_time]&[granularity]
+```
+
+#### Parameters
```
Comment (themiszamani) [inline doc/api/results.md]: add descriptions for 
{report_name}
{group_type}
{lesser_group_type}
{lesser_group_name}
[granularity]

File: doc/api/results.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+Type                  | Description                                                                                               | Required | Default value
+--------------------- | --------------------------------------------------------------------------------------------------------- | -------- | -------------
+`[start_time]`        | UTC time in W3C format                                                                                    | YES      |
+`[end_time]`          | UTC time in W3C format                                                                                    | YES      |
+`[report]`            | Name of the report that contains all the information about the profile, filter tags etc.                  | YES      |
```
Comment (themiszamani) [inline doc/api/results.md]: [report] or  {report_name}?

File: doc/api/results.md [hunk 9]
```diff
@@ -0,0 +1,68 @@
+---
```
Comment (skanct) [inline doc/api/results.md]: Remove this header metadata

File: doc/api/results.md [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+page_title: 'API - Availability & Reliability Results'
+font_title: fa fa-cogs
+description: API calls for retrieving computed Availability and Reliability Results
+---
+
+# API Calls
+
+Name                                                           | Description                                                                                                                                                                                                             | Shortcut
+-------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | -----------------------------
+GET: List Availabilities and Reliabilities for Endpoint Groups | The following methods can be used to obtain Availability and Reliablity metrics per group type elements (i.e. Endpoint Groups, Group of Endpoint Groups etc). Results can be retrieved on daily or monthly granularity. | <a href="#1"> Description</a>
```
Comment (skanct) [inline doc/api/results.md]: Use `[Description](#1)` for the link

File: doc/api/results.md [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+<a id="1"></a>
+
+# GET: List Availabilities and Reliabilities for Endpoint Groups
```
Comment (skanct) [inline doc/api/results.md]: Leave an extra line after each header

File: doc/api/results.md [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+The following methods can be used to obtain Availability and Reliability metrics per group type elements (i.e. Endpoint Groups, Group of Endpoint Groups etc). Results can be retrieved on daily or monthly granularity.
+
+## Endpoint Groups
+### Input
+Endpoint Groups
+
+```
+/results/{report_name}/{lesser_group_type}?[start_time]&[end_time]&[granularity]
```
Comment (skanct) [inline doc/api/results.md]: lesser_group_type????

Comment (makistsantekidis) [inline doc/api/results.md]: should i change it to endpoint_group_type?

File: doc/api/results.md (hunk 7 above)
Comment (skanct) [inline doc/api/results.md]: Change lesser_group_type to something meaningful

File: doc/api/results.md (hunk 6 above)
Comment (skanct) [inline doc/api/results.md]: `application/json` is the correct one. It must be mentioned in the documentation how to ask for xml or json results

File: doc/api/results.md [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+
+### Response
+Headers: `Status: 200 OK`
+
+#### Response body for `/api/v2/results/Report_A/SITE/ST01?start_time=2015-06-20T12:00:00Z&end_time=2015-06-23T23:00:00Z&granularity=daily` API call
+
+```
+<root>
+<group name="GROUP_A" type="GROUP">
+<group name="ST01" type="SITE">
```
Comment (skanct) [inline doc/api/results.md]: Fix identation

File: app/results/pkg_test.go [hunk 14]
```diff
@@ -0,0 +1,398 @@
...
+	endpointGrouAvailabitiyJSON := `{
+   "result": [
+     {
+       "name": "GROUP_A",
+       "type": "GROUP",
+       "endpoints": [
+         {
+           "name": "ST01",
+           "type": "SITE",
+           "availabilities": [
```
Comment (skanct) [inline app/results/pkg_test.go]: Why do we call it availabilities and not results?

File: app/results/Model.go [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+
+type SuperGroup struct {
+	XMLName xml.Name `xml:"group" json:"-"`
+	Name    string   `xml:"name,attr" json:"name"`
+	// SuperGroup    string   `xml:"group,attr" json:"group"`
+	Type          string        `xml:"type,attr" json:"type"`
+	EndpointGroup []interface{} `json:"endpoints"`
+}
+
+//Report struct for formating xml/json
```
Comment (skanct) [inline app/results/Model.go]: Is this needed?

File: app/results/pkg_test.go [hunk 16]
```diff
@@ -0,0 +1,398 @@
...
+					"email":   "g0dk1ng@kingslanding.gov",
+					"api_key": "sansa <3",
+				},
+			}})
+	c.Insert(
+		bson.M{"name": "EGI",
+			"db_conf": []bson.M{
+
+				bson.M{
+					// "store":    "ar",
```
Comment (skanct) [inline app/results/pkg_test.go]: Is this needed?

File: app/results/routing.go [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+	// 	Name("Group_Type").
+	// 	Handler(respond.Respond(...))
+	// TODO: list endpointgroups with type {lgroup_type}
+	// groupTypeSubrouter.
+	// Path("/{group_name}/{lgroup_name}").
+	// Name("EndpointGroup_Type").
+	// Handler(respond.Respond(...))
+
+	// Route for request "api/v2/results/{report_name}/{group_type}/{group_name}"
+	// matches only endpoint groups
```
Comment (skanct) [inline app/results/routing.go]: Is this code needed?

File: app/results/pkg_test.go (hunk 14 above)
Comment (makistsantekidis) [inline app/results/pkg_test.go]: should the top level "results" be renamed "root" and change the "availabilities" to "results" ?

File: app/results/routing.go [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+		Handler(confhandler.Respond(routeGroup))
+	groupSubrouter.
+		Methods("GET").
+		Name("Group Type").
+		Handler(confhandler.Respond(routeGroup))
+
+	// groupSubrouter.
+	// 	Methods("GET").
+	// 	Name("Group Name").
+	// 	Handler(confhandler.Respond(ListSuperGroupResults, "group name"))
```
Comment (skanct) [inline app/results/routing.go]: Is this code needed?

File: doc/api/results.md [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+GET: List Availability and Reliability results for an endpoint group | This method retrieves the results of a specified endpoint group or multiple endpoint groups of a specific type that where computed based on a given report. Results can be retrieved on daily or monthly granularity. | [Description](#1)
+
+<a id="1"></a>
+
+# GET: List Availabilities and Reliabilities for Endpoint Groups
+
+The following methods can be used to obtain a tenant's Availability and Reliability result metrics per Endpoint Group. The api authenticates the tenant using the api-key within the x-api-key header. User can specify time granularity (`monthly` or `daily`) for retrieved results and also format using the `Accept` header. Depending on the form of the request the user can request a single endpoint group results or a bulk of endpoint group results filtered by their type and if necessary their "top-level" group.
```
Comment (skanct) [inline doc/api/results.md]: Please improve a bit this paragraph

File: doc/api/results.md [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+
+## Endpoint Groups
+
+### Input
+
+Endpoint Groups
```
Comment (skanct) [inline doc/api/results.md]: Not needed

File: doc/api/results.md [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+`{endpoint_group_type}` | Type of the the Endpoint Group.                                                                                                     | YES      |
+
+#### Request headers
+
+```
+x-api-key: "tenant_key_value"
+Accept: "application/xml" or "application/json"
+```
+
+### Response
```
Comment (skanct) [inline doc/api/results.md]: Change this to:

```
#### Headers

#### Body
```

File: app/results/pkg_test.go [hunk 22]
```diff
@@ -0,0 +1,398 @@
...
+               "downtime": "0"
+             }
+           ]
+         }
+       ]
+     }
+   ]
+ }`
+	// Check that we must have a 200 ok code
+	suite.Equal(200, response.Code, "Internal Server Error")
```
Comment (skanct) [inline app/results/pkg_test.go]: Really?

File: doc/api/results.md (hunk 19 above)
Comment (makistsantekidis) [inline doc/api/results.md]: Any points to what needs improving?

File: app/results/pkg_test.go (hunk 22 above)
Comment (makistsantekidis) [inline app/results/pkg_test.go]: Yes

Comment (skanct) [inline app/results/pkg_test.go]: What am I missing here? How is it possible to have 200 OK and "Internal Server Error"?

Comment (makistsantekidis) [inline app/results/pkg_test.go]: the function called checks the if the first argument is equal to the second argument and if not, an assertion fails and the  third argument is printed as a message. [testify doc](http://godoc.org/github.com/stretchr/testify/assert)

Comment (skanct) [inline app/results/pkg_test.go]: Can you change the "Internal Server Error" to something meaningful? Like "Incorrect HTTP response code"?

Comment (makistsantekidis) [inline app/results/pkg_test.go]: Yep

File: app/results/pkg_test.go [hunk 23]
```diff
@@ -0,0 +1,396 @@
...
+               "downtime": "0"
+             }
+           ]
+         }
+       ]
+     }
+   ]
+ }`
+
+	fmt.Println(response.Body.String())
```
Comment (dpavlos) [inline app/results/pkg_test.go]: Redundant Println

**Pull request merged**